### PR TITLE
feat: optimize trade for approvals and execution

### PR DESCRIPTION
**Pull request by justindomingue**
- uses smart-order-router when users disable auto router
- chooses SwapRouter based on approval statuses and trade routes
- output of use*Route is now a `@router-sdk/Trade`. Before swapping, we generate a v2/trade, v3/trade or router-sdk/trade
    - that's why there a lot of `V2Trade | V3Trade` simplification to a single `Trade`. It makes the ui layer easier to work with, and the logic layer is eventually consistent to a specific protocol `Trade` object

### CODE REVIEW

File: src/hooks/useApproveCallback.ts [hunk 1]
```diff
@@ -102,15 +134,76 @@ export function useApproveCallback(
...
+ */
+export function useApprovalOptimizedTrade(
+  trade: Trade<Currency, Currency, TradeType> | undefined,
+  allowedSlippage: Percent
+):
+  | V2Trade<Currency, Currency, TradeType>
+  | V3Trade<Currency, Currency, TradeType>
+  | Trade<Currency, Currency, TradeType>
+  | undefined {
+  const hasV2Routes = trade?.routes.some((route) => route.protocol === Protocol.V2)
```
Comment (justindomingue) [inline src/hooks/useApproveCallback.ts:176]: @justindomingue should be `every`

File: src/hooks/useSwapSlippageTolerance.ts [hunk 2]
```diff
@@ -11,31 +11,27 @@ import useGasPrice from './useGasPrice'
...
 
 /**
  * Return a guess of the gas cost used in computing slippage tolerance for a given trade
  * @param trade the trade for which to _guess_ the amount of gas it would cost to execute
  */
-function guesstimateGas(
-  trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType> | undefined
-): number | undefined {
+function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined): number | undefined {
   if (trade instanceof V2Trade) {
```
Comment (justindomingue) [inline src/hooks/useSwapSlippageTolerance.ts:22]: @NoahZinsmeister how would you reconcile this here? We only have a router-sdk/Trade object to wor kwith. Should we stop estimating gas for v2?

File: src/hooks/useApproveCallback.ts [hunk 3]
```diff
@@ -103,23 +135,85 @@ export function useApproveCallback(
...
+
+  const approveCallback = useApproveCallback(
     amountToApprove,
     chainId
       ? trade instanceof V2Trade
         ? V2_ROUTER_ADDRESS[chainId]
         : trade instanceof V3Trade
-        ? v3SwapRouterAddress
-        : undefined
+        ? V3_ROUTER_ADDRESS[chainId]
```
Comment (ianlapham) [inline src/hooks/useApproveCallback.ts:157]: trade will only be a `V3Trade` if its from legeacy router logic?

File: src/hooks/useApproveCallback.ts [hunk 4]
```diff
@@ -103,23 +135,85 @@ export function useApproveCallback(
...
+
+  const optimizedSwapRouter = useMemo(
+    () => getTxOptimizedSwapRouter({ onlyV2Routes, onlyV3Routes, routeHasSplits, approvalStates }),
+    [approvalStates, routeHasSplits, onlyV2Routes, onlyV3Routes]
+  )
+
+  return useMemo(() => {
+    if (!trade) return undefined
+
+    switch (optimizedSwapRouter) {
```
Comment (ianlapham) [inline src/hooks/useApproveCallback.ts:195]: Can we make this `optimizedSwapRouter` enum or soemthing?

File: src/hooks/useAutoRouterSupported.tsx [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { SUPPORTED_CHAINS } from 'state/routing/clientSideSmartOrderRouter/constants'
+
+import { useActiveWeb3React } from './web3'
+
+export default function useAutoRouterSupported(): boolean {
+  const { chainId } = useActiveWeb3React()
+  return Boolean(chainId && SUPPORTED_CHAINS.includes(chainId))
```
Comment (ianlapham) [inline src/hooks/useAutoRouterSupported.tsx:7]: Can we change this to AUTO_ROUTER_SUPPORTED_CHAINS

File: src/hooks/useSwapSlippageTolerance.ts (hunk 2 above)
Comment (ianlapham) [reply src/hooks/useSwapSlippageTolerance.ts:22]: Related - teh auto router only returns gas cost estimate for V3 currently right?

File: src/utils/getTxOptimizedSwapRouter.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import { ApprovalState } from 'hooks/useApproveCallback'
+
+/**
+ * Returns the swap router that will result in the least amount of txs (less gas) for a given swap.
+ * Heuristic:
+ * - if trade contains a single v2-only trade & V2 SwapRouter is approved: use V2 SwapRouter
+ * - if trade contains only v3 & V3 SwapRouter is approved: use V3 SwapRouter
+ * - else: approve and use V2+V3 SwapRouter
+ */
+export function getTxOptimizedSwapRouter({
```
Comment (ianlapham) [inline src/utils/getTxOptimizedSwapRouter.ts:10]: Can we return some enum or something here

**Review by NoahZinsmeister — APPROVED** [review]
looking great, left a few small comments

File: src/hooks/useSwapSlippageTolerance.ts (hunk 2 above)
Comment (justindomingue) [reply src/hooks/useSwapSlippageTolerance.ts:22]: @NoahZinsmeister so basically always run the existing logic we had for v3

@ianlapham possibly yeah in your PR where you carry the gas estimate around we could use it. We do however still fall back to the legacy v3 router some times so we need a solution for that case.

File: src/state/routing/utils.ts [hunk 7]
```diff
@@ -49,22 +52,37 @@ export function computeRoutes(
...
 
+export function transformRoutesToTrade<TTradeType extends TradeType>(
+  route: ReturnType<typeof computeRoutes>,
+  tradeType: TTradeType
+): Trade<Currency, Currency, TTradeType> {
+  return new Trade({
+    v2Routes:
+      route
+        ?.filter((r) => r.routev2 !== null)
+        // typescript does not perform type narrowing from filter
```
[the remark this answers is not in the document]
Comment (justindomingue) [reply src/state/routing/utils.ts:68]: neat the type is kind of verbose but i like type safety 🙏 

```
r is typeof route[0] & { routev2: NonNullable<typeof route[0]['routev2']> }
```

File: src/components/swap/SwapRoute.tsx [hunk 8]
```diff
@@ -28,10 +25,16 @@ export default memo(function SwapRoute({
...
-  trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>
+  trade: Trade<Currency, Currency, TradeType>
   syncing: boolean
 }) {
-  const routingAPIEnabled = useRoutingAPIEnabled()
+  const autoRouterSupported = useAutoRouterSupported()
+
+  const routes = getTokenPath(trade)
+
+  // currently, routes cannot be mixed so while `every` is more correct, some is more efficient
```
[the remark this answers is not in the document]
Comment (justindomingue) [reply src/components/swap/SwapRoute.tsx:35]: yeah you're right, i'll just remove the comment it doesn't add much

File: src/hooks/useApproveCallback.ts (hunk 3 above)
Comment (justindomingue) [reply src/hooks/useApproveCallback.ts:157]: hm in this case actually we pass a v2 | v3 | trade which is derived from the routing api `Trade` by looking at approvals. So if a user has already approved the v3 router, we convert the `Trade` into a `V3Trade`

**Pull request merged**